### ISSUE TITLE
LayoutLMV2: Splitting the document (>512 tokens) into multiples

### ISSUE THREAD

**Issue by PatelKrutarth** [NielsRogge/Transformers-Tutorials#41]
Hello,

**What is the best way to divide the document into multiple parts for LayoutLMv2 without losing any information from the document?**

**What is the best way to devide larger document (tokens > 512) in the custom Dataset class in which `__getitem__` function supposed to return a single entry from the dataset, and each entry/example could be of only upto 512 token?** I am using the custom dataset class for getting and a DataLoader to feed the train and test datasets to the pre-trained model for fine-tuning and testing. In [the response of past issue](https://github.com/NielsRogge/Transformers-Tutorials/issues/23#issuecomment-918030782
), I found that we can create multiple training examples for a given document. In that scenario is it possible to still use the  LayoutLMv2Processor along with the custom dataset to get the document split with multiple subparts? The link of the [tutorial](https://github.com/huggingface/notebooks/blob/master/examples/question_answering.ipynb) mentioned in that response does not answer this specific question. Below is the portion of the custom dataset class that I am using that truncates and pads if needed and does not split the document into multiple parts. 

```
processor = LayoutLMv2Processor.from_pretrained("microsoft/layoutlmv2-base-uncased", revision="no_ocr")

class CustomImageDataset(Dataset):
    def __init__(self, instance_list_path, image_dir, ocr_dir, processor=None, max_length=512):
        self.image_ids = []
        with open(instance_list_path) as ip_file:
            for line in ip_file:
                self.image_ids.append(line.strip())
        self.image_dir = image_dir
        self.ocr_dir = ocr_dir
        self.processor = processor
        self.max_length = max_length

    def __len__(self):
        return len(self.image_ids)

    def __getitem__(self, idx):
        # first, take an image
        curr_id = self.image_ids[idx]
        if os.path.exists(f'{self.image_dir}/{curr_id}.png'):
            image = Image.open(f'{self.image_dir}/{curr_id}.png').convert("RGB")
        else:
            image = Image.open(f'{self.image_dir}/{curr_id}.jpg').convert("RGB")
        image = image.resize((224, 224))

        # get word-level annotations (sorted top to bottom and left to right appearance, basic flow of a document)
        words, boxes, word_labels = self.get_word_box_labels(idx)

        assert len(words) == len(boxes) == len(word_labels)
        
        # use processor to prepare everything
        encoded_inputs = self.processor(image, words, boxes=boxes, word_labels=word_labels, 
                                        padding="max_length", truncation=True, 
                                        return_tensors="pt")
        
        # remove batch dimension
        for k,v in encoded_inputs.items():
          encoded_inputs[k] = v.squeeze()

        assert encoded_inputs.input_ids.shape == torch.Size([512])
        assert encoded_inputs.attention_mask.shape == torch.Size([512])
        assert encoded_inputs.token_type_ids.shape == torch.Size([512])
        assert encoded_inputs.bbox.shape == torch.Size([512, 4])
        assert encoded_inputs.image.shape == torch.Size([3, 224, 224])
        assert encoded_inputs.labels.shape == torch.Size([512]) 
      
        return encoded_inputs
```

**For the processor input, if I manually split words, boxes, and word_labels into small slices and use the exact same entire image as the image input for smaller slices of words, boxes, and word_labels, will that be fine?** I will keep some stride/window to give context information in the continuous blocks. As an example below, if I want to split a document/image containing 200 OCR words input into 2 portions with the first portion containing the tokens corresponding to 150 words with 25 words of overlap between two portions, does that look correct?

```
image = Image.open(f'{self.image_dir}/{curr_id}.png').convert("RGB")

# sorted top to bottom and left to right appearance, basic flow of a document
words, boxes, word_labels = self.get_word_box_labels(idx)
words_1 = words[:150]
words_2 = words[125:]
boxes_1 = boxes[:150]
boxes_2 = boxes[125:]
word_labels_1 = word_labels[:150]
word_labels_2 = word_labels[125:]

# portion-1 with the entire image as the 1st parameter
encoded_inputs_1 = self.processor(image, words_1, boxes=boxes_1, word_labels=word_labels_1, 
                                        padding="max_length", truncation=True, 
                                        return_tensors="pt")

# portion-2 with the entire image as the 1st parameter
encoded_inputs_2 = self.processor(image, words_2, boxes=boxes_2, word_labels=word_labels_2, 
                                        padding="max_length", truncation=True, 
                                        return_tensors="pt")
```

**Moreover, is there any way to use LayoutLMv2Processor for automatically splitting a document into multiple examples without losing any information?** The [reference example](https://github.com/huggingface/notebooks/blob/master/examples/question_answering.ipynb) link that you mentioned in your [response](https://github.com/NielsRogge/Transformers-Tutorials/issues/23#issuecomment-918030782) uses AutoTokenizer and has `return_overflowing_tokens` and `stride` parameters. But, I could not find such parameters for LayoutLMv2Processor. Sometimes, even a small number of words end up with >=512 tokens, and splitting the input based on words can sometimes cause deleting the end of the document using LayoutLMv2Processor with the `padding="max_length"` and `truncation=True` parameters set to it. So, I would like to see if I can still use the LayoutLMv2Processor or the output of it to split the document into multiple parts without losing any information.

**Comment by yellowjs0304** [comment]
@PatelKrutarth Hi, Did you find the good way? I have same issue, when inference, evaluation step with long images

**Comment by purnasai-cyient** [comment]
Hi, @yellowjs0304 @PatelKrutarth I too facing a similar issue. I tried loading the dataset with more than 512 length using stride. like I am doing here:https://github.com/huggingface/transformers/issues/19190. But then i have the issue at train time like i mentioned in the issue.
